photutils/psf/psfphotometry.py

```diff
@@ -0,0 +1,319 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+class PSFPhotometryBase(object):
+    __metaclass__ = abc.ABCMeta
+    
+    @abc.abstractmethod
+    def do_photometry(self):
+        pass
+
+class NStarPSFPhotometry(PSFPhotometryBase):
+    """
+    This is an implementation of the NSTAR algorithm proposed by Stetson
+    (1987) to perform point spread function photometry in crowded fields.
+
+    This class basically implements the loop FIND, GROUP, NSTAR,
+    SUBTRACT, FIND until no more stars are detected.
+    """
+
+    def __init__(self, find, group, bkg, psf, fitter, niters, fitshape):
+        """
+        Attributes
+        ----------
+        find : an instance of any StarFinderBase subclasses
+        group : an instance of any GroupStarsBase subclasses
+        bkg : an instance of any BackgroundBase2D (?) subclasses
+        psf : Fittable2DModel instance
+        fitter : Fitter instance
+        niters : int
+            number of iterations for the loop FIND, GROUP, SUBTRACT, NSTAR
+        fitshape : array-like
+            rectangular shape around the center of a star which will be used
+            to collect the data to do the fitting
+        """
+
+        self.find = find
+        self.group = group
+        self.bkg = bkg
+        self.psf = psf
+        self.fitter = fitter
+        self.niters = niters
+        self.fitshape = fitshape
+
+    @property
+    def find(self):
+        return self._find
+    
+    @find.setter
+    def find(self, find):
+        if isinstance(find, StarFinderBase)
+            self._find = find
+        else:
+            raise ValueError('find is expected to be an instance of '
+                             'StarFinderBase, received {}'.format(type(find)))
+    @property
+    def group(self):
+        return self._group
+
+    @group.setter
+    def group(self, group):
+        if isinstance(group, GroupStarsBase):
+            self._group = group
+        else:
+            raise ValueError('group is expected to be an instance of '
+                             'GroupStarsBase, received {}.'.\
+                             format(type(group)))
+    
+    @property
+    def bkg(self)
```

Comment:
I suggest not to use properties for this here anyway, because it prevents a user for using duck typing.